File: frontend/src/api/applicationBuilder.ts
import { api } from "./client";

export interface Diagnostic {
  code: string;
  severity: "error" | "warning" | "suggestion";
  message: string;
  path: string;
  source: string;
  suggestedFix: string;
  autoFix: boolean;
}

export interface ApplicationProject {
  id: number;
  name: string;
  description: string;
  workflow_id: number | null;
  spec: Record<string, unknown>;
  schema_version: number;
  target: string;
  application_type: string;
  ui_framework: string;
  status: "draft" | "archived";
  created_at: string;
  updated_at: string;
}

export interface SemanticComponent {
  id: string;
  type: string;
  properties?: Record<string, unknown>;
  binding?: string | Record<string, unknown> | null;
  events?: Record<string, unknown>;
  responsive?: Record<string, unknown>;
  locked?: Record<string, boolean>;
  children?: SemanticComponent[];
}

export interface ComponentDefinition {
  type: string; label: string; category: string; container: boolean; defaults: Record<string, unknown>;
  propertySchema: ComponentPropertyDefinition[];
  eventSchema: ComponentEventDefinition[];
  accessibility?: { requiredProperties?: string[]; minimumTouchTarget?: number };
}

export interface ComponentEventDefinition {
  name: string;
  label: string;
  actions: string[];
}

export interface ComponentPropertyDefinition {
  key: string;
  label: string;
  type: "string" | "multiline" | "boolean" | "number" | "enum" | "json" | "responsive-columns" | "table-columns" | "chart-series";
  required?: boolean;
  options?: string[];
  minimum?: number;
  maximum?: number;
  breakpoints?: string[];
  maximumItems?: number;
  columnTypes?: string[];
  tones?: string[];
}

export interface DesignPresetDefinition {
  id: string; label: string; description: string; tokens: Record<string, string>;
}

export interface DesignTemplateDefinition {
  id: string; label: string; description: string; category?: string; root: SemanticComponent;
  parameters: DesignTemplateParameterDefinition[];
}

export interface DesignTemplateParameterDefinition {
  key: string;
  label: string;
  type: "string" | "number" | "boolean" | "enum";
  default: string | number | boolean;
  required?: boolean;
  maximumLength?: number;
  minimum?: number;
  maximum?: number;
  options?: string[];
  targets: Array<{ componentId: string; property: string }>;
}

export interface SemanticComponentCatalog {
  schemaVersion: number;
  components: ComponentDefinition[];
  designTokens: Record<string, string[]>;
  bindingSources: string[];
  bindingDefinitions: Array<{ id: string; label: string; referenceLabel: string }>;
  eventActions: Array<{ id: string; label: string; targetLabel: string; targetSection: string | null }>;
  accessibilityAudit: { minimumContrast: number; minimumLargeTextContrast: number; minimumTouchTarget: number; minimumFocusIndicator: number };
  presets: DesignPresetDefinition[];
  composites: DesignTemplateDefinition[];
  patterns: DesignTemplateDefinition[];
  previewStates: Array<{ id: "default" | "loading" | "empty" | "error" | "disabled"; label: string; description: string }>;
}

export interface ApplicationSchemaCatalog {
  schemaVersion: number;
  semanticComponents: SemanticComponentCatalog;
}

export type ApplicationEntityFieldType = "string" | "integer" | "number" | "boolean" | "datetime" | "json";
export interface ApplicationEntityField {
  id: string; type: ApplicationEntityFieldType; nullable?: boolean; default?: unknown; hasDefault?: boolean;
  maxLength?: number | null; unique?: boolean; indexed?: boolean;
  reference?: { entityId: string; onDelete: "restrict" | "cascade" | "set-null" } | null;
}
export interface ApplicationEntity {
  id: string; displayName?: string; tableName?: string | null; fields: ApplicationEntityField[];
  crud?: { enabled: boolean; operations: Array<"create" | "read" | "list" | "update" | "delete">; basePath?: string | null };
}

export type ApplicationClientStateType = "string" | "integer" | "number" | "boolean" | "object" | "array";
export interface ApplicationClientState {
  id: string;
  type: ApplicationClientStateType;
  initialValue: unknown;
  nullable?: boolean;
}

export interface ApplicationQuery {
  id: string;
  source: "entity" | "api";
  entityId?: string | null;
  endpointId?: string | null;
  input?: Record<string, unknown>;
  resultPath?: string;
  filters?: Array<{ field: string; operator: "eq" | "ne" | "contains" | "starts-with" | "gt" | "gte" | "lt" | "lte" | "is-null"; value?: unknown }>;
  sort?: Array<{ field: string; direction: "asc" | "desc" }>;
  pagination?: "none" | "offset";
  limit: number;
  autoLoad: boolean;
  cachePolicy: "network-only" | "memory";
  staleTimeSeconds: number;
}

export interface ApplicationPatchOperation {
  op: "add" | "remove" | "replace" | "move";
  path: string;
  from?: string;
  value?: unknown;
}

export interface ApplicationPatchPreview {
  valid: boolean;
  baseChecksum: string;
  resultChecksum: string;
  patchedSpec: Record<string, unknown>;
  appliedPatches: ApplicationPatchOperation[];
  diagnostics: Diagnostic[];
}

export interface ApplicationDesignProposal {
  id: string;
  direction: "simple" | "balanced" | "dense";
  title: string;
  summary: string;
  rationale: string[];
  patches: ApplicationPatchOperation[];
  warnings: string[];
  preview: ApplicationPatchPreview;
}

export interface LlmEndpoint {
  base_url: string;
  models: string[];
  managed?: boolean;
  selected?: boolean;
}

export interface FrameworkCapability {
  id: string;
  label: string;
  language: string;
  platforms: string[];
  status: string;
  source: boolean;
  build: boolean;
  package: boolean;
  phase: string;
  details: { sdks: string[]; features: Record<string, boolean> };
  matrix: PlatformCapabilityMatrix;
}

export interface PlatformCapabilityMatrix {
  spec: string; source: string; localBuild: string; remoteBuild: string;
  package: string; signing: string; store: string; stability: string;
}

export interface CapabilityCatalog {
  phase: string;
  generationAvailable: boolean;
  buildAvailable: boolean;
  build: { available: boolean; sdk: string; sdkPath: string | null; systemdUser: boolean; network: "denied"; maxConcurrent: number; note: string };
  frameworks: FrameworkCapability[];
  nodes: Array<{ type: string; targets: Record<string, { support: string; reason: string }> }>;
  host: { os: string; architecture: string; sdks: Record<string, boolean>; note: string };
}

export interface PlatformAdvisorRequest {
  platforms: Array<"web" | "linux" | "windows" | "macos" | "android" | "ios">;
  offline: boolean; localFiles: boolean; tray: boolean; background: boolean; gpu: boolean;
  embeddedServer: boolean; store: boolean; preferredLanguage: "any" | "csharp" | "typescript" | "rust" | "dart" | "kotlin" | "cpp";
  preferNativeFeel: boolean; preferWebReuse: boolean; preferSmallSize: boolean;
}

export interface PlatformRecommendation {
  frameworkId: string; label: string; score: number; platforms: string[]; language: string; status: string;
  reasons: string[]; constraints: string[]; matrix: PlatformCapabilityMatrix;
}

export interface PlatformAdvisorResult {
  phase: string; recommendedId: string; requestedPlatforms: string[];
  recommendations: PlatformRecommendation[]; host: CapabilityCatalog["host"]; note: string;
}

export interface ApplicationPreflightResult {
  phase: string; validSpec: boolean; readyForGeneration: boolean; readyForLocalBuild: boolean;
  targets: Array<{ id: string; frameworkId: string; label: string; platforms: string[]; matrix: PlatformCapabilityMatrix; requiredSdks: string[]; missingSdks: string[] }>;
  diagnostics: Diagnostic[]; host: CapabilityCatalog["host"];
  sideEffects: { executor: false; network: false; subprocess: false; filesystemWrite: false; secretResolution: false };
}

export interface ApplicationSourcePreview {
  phase: string; ready: boolean; diagnostics: Diagnostic[];
  sideEffects: { executor: false; network: false; subprocess: false; filesystemWrite: false; secretResolution: false };
  generator?: { id: string; version: string }; deterministic?: boolean;
  archiveName?: string; archiveChecksum?: string; sourceChecksum?: string; archiveBytes?: number;
  files?: Array<{ path: string; sha256: string; bytes: number; kind: "managed" | "extension" | "config" | "manifest" }>;
  manifest?: { input: { specChecksum: string; workflowChecksum: string; targetId: string; framework: string }; managedFiles: string[]; extensionFiles: string[]; configFiles: string[] };
}

export interface ValidationResult {
  valid: boolean;
  normalizedSpec: Record<string, unknown>;
  workflowIr: null | { inputs: unknown[]; outputs: unknown[]; nodes: unknown[]; edges: unknown[]; capabilities: string[]; side_effects: string[] };
  applicationIr: { pages: unknown[]; entities: unknown[]; api_endpoints: unknown[]; targets: unknown[] };
  diagnostics: Diagnostic[];
  capability: { target: string; generationAvailable: boolean; buildAvailable: boolean; note: string };
}

export interface ApplicationBuildArtifact {
  id: number; path: string; kind: "source" | "binary" | string; mimeType: string; size: number; checksum: string;
}

export interface ApplicationBuild {
  id: number; projectId: number; targetId: string; framework: string;
  status: "queued" | "preparing" | "generating" | "restoring" | "building" | "testing" | "canceling" | "completed" | "failed" | "cancelled" | "timed_out" | "interrupted";
  sourceChecksum: string; archiveChecksum: string; generator: { id?: string; version?: string }; sdk: string;
  timeoutSeconds: number; result: string; exitCode: number | null; error: string;
  createdAt: string; startedAt: string | null; finishedAt: string | null;
  isolation: { systemdUser: true; network: "denied"; memoryMax: string; tasksMax: number; cpuQuota: string };
  artifacts: ApplicationBuildArtifact[]; logs?: string;
}

export const applicationBuilderApi = {
  list: (workflowId?: number) => api<ApplicationProject[]>(`/application-projects${workflowId ? `?workflow_id=${workflowId}` : ""}`),
  get: (id: number) => api<ApplicationProject>(`/application-projects/${id}`),
  create: (body: { name: string; description?: string; workflow_id?: number }) =>
    api<ApplicationProject>("/application-projects", { method: "POST", json: body }),
  createFromWorkflow: (workflowId: number, body: { source: "draft" | "published"; name?: string }) =>
    api<ApplicationProject>(`/workflows/${workflowId}/application-projects`, { method: "POST", json: body }),
  update: (id: number, body: { name?: string; description?: string; spec?: Record<string, unknown> }) =>
    api<ApplicationProject>(`/application-projects/${id}`, { method: "PATCH", json: body }),
  remove: (id: number) => api<void>(`/application-projects/${id}`, { method: "DELETE" }),
  capabilities: () => api<CapabilityCatalog>("/application-builder/capabilities"),
  advisePlatforms: (body: PlatformAdvisorRequest) => api<PlatformAdvisorResult>("/application-builder/platform-advisor", { method: "POST", json: body }),
  preflight: (project: ApplicationProject, spec: Record<string, unknown>) => api<ApplicationPreflightResult>("/application-builder/preflight", {
    method: "POST", json: { spec, workflow_id: project.workflow_id },
  }),
  sourcePreview: (projectId: number, targetId: string) => api<ApplicationSourcePreview>(`/application-projects/${projectId}/source-preview?target_id=${encodeURIComponent(targetId)}`),
  downloadSource: async (projectId: number, targetId: string) => {
    const response = await fetch(`/api/v1/application-projects/${projectId}/source-archive`, {
      method: "POST", credentials: "same-origin",
      headers: { "Content-Type": "application/json", "X-Requested-With": "ControlDeck" },
      body: JSON.stringify({ targetId }),
    });
    if (!response.ok) {
      let message = `Source生成に失敗しました (${response.status})`;
      try {
        const payload = await response.json() as { detail?: string | { diagnostics?: Diagnostic[] } };
        if (typeof payload.detail === "string") message = payload.detail;
        else if (payload.detail?.diagnostics?.length) message = payload.detail.diagnostics.map((item) => `${item.code}: ${item.message}`).join(" / ");
      } catch { /* non-JSON error */ }
      throw new Error(message);
    }
    const disposition = response.headers.get("Content-Disposition") ?? "";
    const filename = disposition.match(/filename="([^"]+)"/)?.[1] ?? "generated-source.zip";
    return { blob: await response.blob(), filename, checksum: response.headers.get("X-ControlDeck-Source-SHA256") ?? "" };
  },
  builds: (projectId: number) => api<ApplicationBuild[]>(`/application-projects/${projectId}/builds`),
  startBuild: (projectId: number, targetId: string, timeoutSeconds = 900) =>
    api<ApplicationBuild>(`/application-projects/${projectId}/builds`, { method: "POST", json: { targetId, timeoutSeconds } }),
  buildLogs: (buildId: number) => api<ApplicationBuild>(`/application-builds/${buildId}/logs`),
  cancelBuild: (buildId: number) => api<ApplicationBuild>(`/application-builds/${buildId}/cancel`, { method: "POST" }),
  removeBuild: (buildId: number) => api<void>(`/application-builds/${buildId}`, { method: "DELETE" }),
  artifactUrl: (buildId: number, artifactId: number) => `/api/v1/application-builds/${buildId}/artifacts/${artifactId}`,
  schema: () => api<ApplicationSchemaCatalog>("/application-builder/schema"),
  validateSpec: (project: ApplicationProject, spec: Record<string, unknown>) => api<ValidationResult>("/application-builder/validate", {
    method: "POST", json: { spec, workflow_id: project.workflow_id, target: project.target },
  }),
  validate: (project: ApplicationProject) => api<ValidationResult>("/application-builder/validate", {
    method: "POST", json: { spec: project.spec, workflow_id: project.workflow_id, target: project.target },
  }),
  previewPatches: (spec: Record<string, unknown>, patches: ApplicationPatchOperation[]) =>
    api<ApplicationPatchPreview>("/application-builder/patches/preview", { method: "POST", json: { spec, patches } }),
  applyPatches: (projectId: number, baseChecksum: string, patches: ApplicationPatchOperation[]) =>
    api<{ project: ApplicationProject; patch: ApplicationPatchPreview }>(`/application-projects/${projectId}/patches/apply`, {
      method: "POST", json: { base_checksum: baseChecksum, patches },
    }),
  llmEndpoints: () => api<LlmEndpoint[]>("/workflows/llm-endpoints"),
  designProposals: (projectId: number, body: {
    instruction: string; scope: "application" | "page" | "component" | "mobile";
    target_id?: string; mode: "preserve" | "balanced" | "redesign"; base_url: string; model: string;
  }) => api<{ proposals: ApplicationDesignProposal[] }>(`/application-projects/${projectId}/design-proposals`, { method: "POST", json: body }),
};

File: backend/app/schemas/application_builder.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class ApplicationProjectCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    description: str = Field(default="", max_length=4000)
    workflow_id: int | None = Field(default=None, ge=1)
    spec: dict[str, Any] | None = None


class ApplicationProjectUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=4000)
    spec: dict[str, Any] | None = None


class ApplicationValidateBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    spec: dict[str, Any]
    workflow_id: int | None = Field(default=None, ge=1)
    workflow_version_id: int | None = Field(default=None, ge=1)
    target: str = Field(default="csharp", pattern="^(csharp|cpp)$")


class PlatformAdvisorRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    platforms: list[Literal["web", "linux", "windows", "macos", "android", "ios"]] = Field(default_factory=lambda: ["web"], min_length=1, max_length=6)
    offline: bool = False
    local_files: bool = Field(default=False, alias="localFiles")
    tray: bool = False
    background: bool = False
    gpu: bool = False
    embedded_server: bool = Field(default=False, alias="embeddedServer")
    store: bool = False
    preferred_language: Literal["any", "csharp", "typescript", "rust", "dart", "kotlin", "cpp"] = Field(default="any", alias="preferredLanguage")
    prefer_native_feel: bool = Field(default=False, alias="preferNativeFeel")
    prefer_web_reuse: bool = Field(default=False, alias="preferWebReuse")
    prefer_small_size: bool = Field(default=False, alias="preferSmallSize")


class ApplicationPreflightBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    spec: dict[str, Any]
    workflow_id: int | None = Field(default=None, ge=1)
    workflow_version_id: int | None = Field(default=None, ge=1)


class ApplicationSourceRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    target_id: str = Field(alias="targetId", min_length=1, max_length=128, pattern=r"^[A-Za-z0-9_.-]+$")


class ApplicationBuildRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    target_id: str = Field(alias="targetId", min_length=1, max_length=128, pattern=r"^[A-Za-z0-9_.-]+$")
    timeout_seconds: int = Field(default=900, alias="timeoutSeconds", ge=60, le=3600)


class ApplicationPatchOperation(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    op: Literal["add", "remove", "replace", "move"]
    path: str = Field(min_length=1, max_length=2048)
    from_path: str | None = Field(default=None, alias="from", max_length=2048)
    value: Any = None


class ApplicationPatchPreviewBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    spec: dict[str, Any]
    patches: list[ApplicationPatchOperation] = Field(min_length=1, max_length=200)


class ApplicationPatchApplyBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    base_checksum: str = Field(min_length=64, max_length=64, pattern=r"^[0-9a-f]{64}$")
    patches: list[ApplicationPatchOperation] = Field(min_length=1, max_length=200)


class ApplicationDesignProposalRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    instruction: str = Field(min_length=3, max_length=4000)
    scope: Literal["application", "page", "component", "mobile"] = "application"
    target_id: str | None = Field(default=None, max_length=128)
    mode: Literal["preserve", "balanced", "redesign"] = "balanced"
    base_url: str = Field(min_length=8, max_length=512)
    model: str = Field(min_length=1, max_length=256)


class ApplicationDesignProposal(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1, max_length=64)
    direction: Literal["simple", "balanced", "dense"]
    title: str = Field(min_length=1, max_length=120)
    summary: str = Field(min_length=1, max_length=1000)
    rationale: list[str] = Field(default_factory=list, max_length=8)
    patches: list[ApplicationPatchOperation] = Field(min_length=1, max_length=200)
    warnings: list[str] = Field(default_factory=list, max_length=8)


class ApplicationDesignProposalEnvelope(BaseModel):
    model_config = ConfigDict(extra="forbid")

    proposals: list[ApplicationDesignProposal] = Field(min_length=3, max_length=3)


class WorkflowApplicationCreate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=4000)
    source: str = Field(default="draft", pattern="^(draft|published)$")


class ApplicationInfoV1(BaseModel):
    model_config = ConfigDict(extra="allow")

    name: str = Field(pattern=r"^[A-Za-z][A-Za-z0-9_.-]{0,127}$")
    displayName: str = ""
    description: str = ""
    applicationType: str = "web"
    authentication: str = "local"
    database: str = "none"


class TargetProfileV1(BaseModel):
    model_config = ConfigDict(extra="allow")

    id: str
    platforms: list[str]
    framework: str


class LlmRuntimeV1(BaseModel):
    model_config = ConfigDict(extra="allow")

    mode: Literal["none", "external", "embedded", "remote"] = "none"
    provider: Literal["ollama", "lmstudio", "openai-compatible", "controldeck"] | None = None
    bundleRuntime: bool = False
    baseUrlEnvironment: str = "LLM_BASE_URL"
    modelEnvironment: str = "LLM_MODEL"


class ComponentLockV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    structure: bool = False
    binding: bool = False
    style: bool = False
    position: bool = False
    content: bool = False


class SemanticComponentV1(BaseModel):
    """Framework名を含まない、generator間で共有するUI部品。"""

    model_config = ConfigDict(extra="allow")

    id: str = Field(min_length=1, max_length=128)
    type: str = Field(min_length=1, max_length=128)
    properties: dict[str, Any] = Field(default_factory=dict)
    binding: str | dict[str, Any] | None = None
    events: dict[str, Any] = Field(default_factory=dict)
    responsive: dict[str, Any] = Field(default_factory=dict)
    locked: ComponentLockV1 = Field(default_factory=ComponentLockV1)
    children: list["SemanticComponentV1"] = Field(default_factory=list)


class ApplicationPageV1(BaseModel):
    model_config = ConfigDict(extra="allow")

    id: str = Field(min_length=1, max_length=128)
    title: str = ""
    description: str = ""
    root: SemanticComponentV1 | None = None


class ApplicationApiEndpointV1(BaseModel):
    """Framework非依存のHTTP endpoint。handler codeは保存せずWorkflowへ結び付ける。"""

    model_config = ConfigDict(extra="allow")

    id: str = Field(min_length=1, max_length=128, pattern=r"^[A-Za-z][A-Za-z0-9_.-]{0,127}$")
    method: Literal["POST"] = "POST"
    path: str = Field(
        min_length=1, max_length=512,
        pattern=r"^/(?:$|(?:[A-Za-z0-9._~-]+|\{[A-Za-z][A-Za-z0-9_]*\})(?:/(?:[A-Za-z0-9._~-]+|\{[A-Za-z][A-Za-z0-9_]*\}))*)$",
    )
    workflowId: int = Field(ge=1)
    mode: Literal["sync", "async"] = "sync"
    authentication: Literal["inherit", "anonymous"] = "inherit"
    requestSchema: dict[str, Any] = Field(default_factory=dict)
    responseSchema: dict[str, Any] = Field(default_factory=dict)
    timeoutSeconds: float = Field(default=120, ge=0.1, le=7200)


class ApplicationBackgroundJobV1(BaseModel):
    """Phase Cで生成するbackground jobの宣言。実行内容はWorkflowだけを参照する。"""

    model_config = ConfigDict(extra="allow")

    id: str = Field(min_length=1, max_length=128, pattern=r"^[A-Za-z][A-Za-z0-9_.-]{0,127}$")
    workflowId: int = Field(ge=1)
    trigger: Literal["manual", "interval", "daily", "cron"] = "manual"
    schedule: str = Field(default="", max_length=256)
    timeZone: str = Field(default="UTC", min_length=1, max_length=64)
    input: dict[str, Any] = Field(default_factory=dict)
    enabled: bool = True
    timeoutSeconds: float = Field(default=7200, ge=0.1, le=7200)
    concurrencyPolicy: Literal["skip", "queue-one"] = "skip"
    catchUpPolicy: Literal["skip", "run-once"] = "run-once"


class ApplicationClientStateV1(BaseModel):
    """Browser-memory state declared independently from component bindings/events."""

    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1, max_length=128, pattern=r"^[A-Za-z][A-Za-z0-9_-]{0,127}$")
    type: Literal["string", "integer", "number", "boolean", "object", "array"]
    initialValue: Any
    nullable: bool = False


class ApplicationQueryFilterV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    field: str = Field(min_length=1, max_length=128, pattern=r"^[A-Za-z][A-Za-z0-9_]{0,127}$")
    operator: Literal["eq", "ne", "contains", "starts-with", "gt", "gte", "lt", "lte", "is-null"]
    value: Any = None


class ApplicationQuerySortV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    field: str = Field(min_length=1, max_length=128, pattern=r"^[A-Za-z][A-Za-z0-9_]{0,127}$")
    direction: Literal["asc", "desc"] = "asc"


class ApplicationQueryV1(BaseModel):
    """Typed, read-only collection query executed by the generated browser UI."""

    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1, max_length=128, pattern=r"^[A-Za-z][A-Za-z0-9_-]{0,127}$")
    source: Literal["entity", "api"] = "entity"
    entityId: str | None = Field(default=None, min_length=1, max_length=128, pattern=r"^[A-Za-z][A-Za-z0-9_]{0,127}$")
    endpointId: str | None = Field(default=None, min_length=1, max_length=128, pattern=r"^[A-Za-z][A-Za-z0-9_.-]{0,127}$")
    input: dict[str, Any] = Field(default_factory=dict)
    resultPath: str = Field(default="", max_length=256, pattern=r"^(?:[A-Za-z][A-Za-z0-9_]*(?:\.[A-Za-z][A-Za-z0-9_]*)*)?$")
    filters: list[ApplicationQueryFilterV1] = Field(default_factory=list, max_length=20)
    sort: list[ApplicationQuerySortV1] = Field(default_factory=list, max_length=3)
    pagination: Literal["none", "offset"] = "offset"
    limit: int = Field(default=20, ge=1, le=100)
    autoLoad: bool = True
    cachePolicy: Literal["network-only", "memory"] = "memory"
    staleTimeSeconds: int = Field(default=30, ge=0, le=3600)


class ApplicationEntityReferenceV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    entityId: str = Field(min_length=1, max_length=128, pattern=r"^[A-Za-z][A-Za-z0-9_]{0,127}$")
    onDelete: Literal["restrict", "cascade", "set-null"] = "restrict"


class ApplicationEntityFieldV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1, max_length=128, pattern=r"^[A-Za-z][A-Za-z0-9_]{0,127}$")
    type: Literal["string", "integer", "number", "boolean", "datetime", "json"]
    nullable: bool = False
    default: Any = None
    hasDefault: bool = False
    maxLength: int | None = Field(default=None, ge=1, le=1_000_000)
    unique: bool = False
    indexed: bool = False
    reference: ApplicationEntityReferenceV1 | None = None


class ApplicationEntityCrudV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    enabled: bool = False
    operations: list[Literal["create", "read", "list", "update", "delete"]] = Field(
        default_factory=lambda: ["create", "read", "list", "update", "delete"],
        min_length=1, max_length=5,
    )
    basePath: str | None = Field(
        default=None, min_length=1, max_length=256,
        pattern=r"^/api/[A-Za-z][A-Za-z0-9_-]*(?:/[A-Za-z][A-Za-z0-9_-]*)*$",
    )


class ApplicationEntityV1(BaseModel):
    """SQLite-backed entity. The generated runtime owns id/createdAt/updatedAt columns."""

    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1, max_length=128, pattern=r"^[A-Za-z][A-Za-z0-9_]{0,127}$")
    displayName: str = Field(default="", max_length=256)
    tableName: str | None = Field(default=None, min_length=1, max_length=128, pattern=r"^[a-z][a-z0-9_]{0,127}$")
    fields: list[ApplicationEntityFieldV1] = Field(min_length=1, max_length=100)
    crud: ApplicationEntityCrudV1 = Field(default_factory=ApplicationEntityCrudV1)


class ApplicationSpecV1(BaseModel):
    """未知fieldを許容し、後続schema versionの情報をround-tripで保持する。"""

    model_config = ConfigDict(extra="allow")

    schemaVersion: int = Field(default=1, ge=1)
    application: ApplicationInfoV1
    theme: dict[str, Any] = Field(default_factory=dict)
    navigation: dict[str, Any] = Field(default_factory=dict)
    pages: list[ApplicationPageV1] = Field(default_factory=list)
    entities: list[ApplicationEntityV1] = Field(default_factory=list, max_length=100)
    apiEndpoints: list[ApplicationApiEndpointV1] = Field(default_factory=list)
    backgroundJobs: list[ApplicationBackgroundJobV1] = Field(default_factory=list)
    clientState: list[ApplicationClientStateV1] = Field(default_factory=list, max_length=100)
    queries: list[ApplicationQueryV1] = Field(default_factory=list, max_length=100)
    workflows: list[dict[str, Any]] = Field(default_factory=list)
    permissions: list[Any] = Field(default_factory=list)
    targets: list[TargetProfileV1] = Field(default_factory=list)
    llmRuntime: LlmRuntimeV1 = Field(default_factory=LlmRuntimeV1)
